projects/fal/src/fal/api.py

```diff
@@ -943,12 +955,104 @@ def function(
+    return False
+
+
+@dataclass
+class BundlePath:
+    path: str
+    filter: Callable[[str], bool] = field(default=INCLUDE_PYTHON_FILES)
+
+    def __iter__(self) -> Iterator[str]:
+        if os.path.isdir(self.path):
+            for root, _dirs, files in os.walk(self.path):
+                for file in files:
+                    if self.filter(os.path.join(root, file)):
+                        yield os.path.join(root, file)
+        else:
+            if self.filter(self.path):
+                yield self.path
+
+
+@dataclass
+class Bundle:
+    zipfile_bytes: bytes
+    _zipfile: ZipFile | None = None
+
+    def __getstate__(self) -> dict[str, Any]:
+        return {
+            "zipfile_bytes": self.zipfile_bytes,
+        }
+
+    def __setstate__(self, state: dict[str, Any]):
+        self.zipfile_bytes = state["zipfile_bytes"]
+
+    @classmethod
+    def _add_file(cls, zipfile: ZipFile, path: str, arcname: str):
+        size = os.path.getsize(path)
+        if size > MAX_BUNDLE_FILE_SIZE:
+            raise ValueError(
+                f"Bundle file {arcname} is larger than {MAX_BUNDLE_FILE_SIZE} bytes"
+            )
+        zipfile.write(path, arcname)
+
+    @classmethod
+    def from_paths(cls, root: str, paths: list[str | BundlePath]) -> Bundle:
+        buffer = io.BytesIO()
+
+        with ZipFile(buffer, "w") as zipfile:
+            for path in paths:
+                if isinstance(path, BundlePath):
+                    bundle_path = path
+                    str_path = path.path
+                else:
+                    bundle_path = BundlePath(path)
+                    str_path = path
+
+                if not os.path.isabs(str_path):
+                    str_path = os.path.join(root, str_path)
+
+                norm_path = os.path.normpath(str_path)
+                if not norm_path.startswith(root):
+                    raise ValueError(
+                        f"Bundle path {path} is outside the fal.App directory {root}"
+                    )
+
+                bundle_path = BundlePath(norm_path, filter=bundle_path.filter)
+                for file in bundle_path:
+                    arcname = os.path.relpath(file, root)
+                    cls._add_file(zipfile, file, arcname)
+
+        zipfile_bytes = buffer.getvalue()
+        size = len(zipfile_bytes)
+        if size > MAX_BUNDLE_TOTAL_SIZE:
+            raise ValueError(
+                f"Bundle size is too big ({size} > {MAX_BUNDLE_TOTAL_SIZE})"
+            )
+
+        return cls(zipfile_bytes)
+
+    def unpack(self, path: str):
+        buffer = io.BytesIO(self.zipfile_bytes)
+        with ZipFile(buffer, "r") as zipfile:
+            zipfile.extractall(path)
+
+
 # implementation
 def function(  # type: ignore
     kind: str = "virtualenv",
     *,
     host: Host | None = None,
     local_python_modules: list[str] | None = None,
+    bundle_paths: list[str | BundlePath] | None = None,
```

Comment:
```suggestion
    bundle_paths: list[str] | None = None,
    bundle_paths_ignore: list[str] | None = None,
```

some feedback from discussing other features with @Warlord-K , maybe we should just keep it simple and add a simple `_ignore` instead of `BundlePath`.